File: src/Classes/GrowthPush/android/GrowthPushPlugin.cpp
//
//  GrowthPushPlugin.cpp
//  GrowthPushPlugin
//
//  Created by TSURUDA Ryo on 2013/12/08.
//  Copyright (c) 2013年 TSURUDA Ryo. All rights reserved.
//

#include "GrowthPushPlugin.h"

#include "cocos2d.h"
#include "platform/android/jni/JniHelper.h"
#include <jni.h>

USING_NS_CC;
USING_NS_GROWTHPUSH;

const char *JavaClassName = "com/growthpush/cocos2dx/GrowthPushPlugin";

GrowthPushPlugin::GrowthPushPlugin(void)
{
}

GrowthPushPlugin::~GrowthPushPlugin(void)
{
}

void GrowthPushPlugin::initialize(int applicationId, const char *secret, growthpush::GPEnvironment environment, bool debug)
{
    CCAssert(secret, "secret should not be NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "initialize", "(ILjava/lang/String;IZ)V")) {
        jstring jSecret = t.env->NewStringUTF(secret);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, applicationId, jSecret, environment, debug);
        t.env->DeleteLocalRef(jSecret);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::registerDeviceToken(const char *senderId)
{
    CCAssert(senderId, "senderId should not be NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "register", "(Ljava/lang/String;)V")) {
        jstring jSenderId = t.env->NewStringUTF(senderId);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jSenderId);
        t.env->DeleteLocalRef(jSenderId);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::trackEvent(const char *name)
{
    CCAssert(name, "name should not be NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "trackEvent", "(Ljava/lang/String;)V")) {
        jstring jName = t.env->NewStringUTF(name);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jName);
        t.env->DeleteLocalRef(jName);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::trackEvent(const char *name, const char *value)
{
    CCAssert(name, "name should not be  NULL");
    CCAssert(value, "value should not be  NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "trackEvent", "(Ljava/lang/String;Ljava/lang/String;)V")) {
        jstring jName = t.env->NewStringUTF(name);
        jstring jValue = t.env->NewStringUTF(value);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jName, jValue);
        t.env->DeleteLocalRef(jName);
        t.env->DeleteLocalRef(jValue);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::setTag(const char *name)
{
    CCAssert(name, "name should not be  NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "setTag", "(Ljava/lang/String;)V")) {
        jstring jName = t.env->NewStringUTF(name);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jName);
        t.env->DeleteLocalRef(jName);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::setTag(const char *name, const char *value)
{
    CCAssert(name, "name should not be  NULL");
    CCAssert(value, "value should not be  NULL");
    
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "setTag", "(Ljava/lang/String;Ljava/lang/String;)V")) {
        jstring jName = t.env->NewStringUTF(name);
        jstring jValue = t.env->NewStringUTF(value);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jName, jValue);
        t.env->DeleteLocalRef(jName);
        t.env->DeleteLocalRef(jValue);
        t.env->DeleteLocalRef(t.classID);
    }
}

void GrowthPushPlugin::setDeviceTags(void)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t, JavaClassName, "setDeviceTags", "()V")) {
        t.env->CallStaticVoidMethod(t.classID, t.methodID);
        t.env->DeleteLocalRef(t.classID);
    }
}

File: src/proj.android/src/com/growthpush/cocos2dx/GrowthPushPlugin.java
package com.growthpush.cocos2dx;

import android.app.Activity;
import android.content.Context;

import com.growthpush.GrowthPush;
import com.growthpush.model.Environment;

public class GrowthPushPlugin {
    private static Context mContext = null;
        
    public GrowthPushPlugin(Activity activity) {
        mContext = activity.getApplicationContext();
    }
    
    public static void initialize(int applicationId, final String secret) {
        GrowthPush.getInstance().initialize(mContext, applicationId, secret);
    }
    
    public static void initialize(int applicationId, final String secret, int environment) {
        GrowthPush.getInstance().initialize(mContext, applicationId, secret, getEnvironmentFromCocos(environment));
    }
    
    public static void initialize(int applicationId, final String secret, int environment, boolean debug) {
        GrowthPush.getInstance().initialize(mContext, applicationId, secret, getEnvironmentFromCocos(environment), debug);
    }
    
    public static void register(final String senderId) {
        GrowthPush.getInstance().register(senderId);
    }
    
    public static void trackEvent(final String name) {
        GrowthPush.getInstance().trackEvent(name);
    }
    
    public static void trackEvent(final String name, final String value) {
        GrowthPush.getInstance().trackEvent(name, value);
    }
    
    public static void setTag(final String name) {
        GrowthPush.getInstance().setTag(name);
    }
    
    public static void setTag(final String name, final String value) {
        GrowthPush.getInstance().setTag(name, value);
    }
    
    public static void setDeviceTags() {
        GrowthPush.getInstance().setDeviceTags();
    }
    
    /*
     Convert com.growthpush.model.Environment from GPEnvironment value of Cocos2d-x
     
     @param environment GPEnvironment value of Cocos2d-x
     @return Environment value of Java
     */
    private static Environment getEnvironmentFromCocos(int environment) {
        if (environment == Environment.development.ordinal()) {
            return Environment.development;
        } else if(environment == Environment.production.ordinal()) {
            return Environment.production;
        }
        throw new IllegalArgumentException("Unknown environment");
    }
}
